lib/nrf_modem_lib/trace_backends/uart/uart.c

```diff
@@ -1,142 +1,172 @@
 /*
- * Copyright (c) 2022 Nordic Semiconductor ASA
+ * Copyright (c) 2022-2023 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
  */
 
-#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/uart.h>
 #include <zephyr/kernel.h>
 #include <zephyr/logging/log.h>
 #include <modem/trace_backend.h>
-#include <nrfx_uarte.h>
 
 LOG_MODULE_REGISTER(modem_trace_backend, CONFIG_MODEM_TRACE_BACKEND_LOG_LEVEL);
 
-#define UART1_NL DT_NODELABEL(uart1)
-PINCTRL_DT_DEFINE(UART1_NL);
-static const nrfx_uarte_t uarte_inst = NRFX_UARTE_INSTANCE(1);
+#define UART_DEVICE_NODE DT_CHOSEN(nordic_modem_trace_uart)
+#define TRACE_UART_BAUD_REQUIRED 1000000
 
-/* Maximum time to wait for a UART transfer to complete before giving up. */
-#define UART_TX_WAIT_TIME_MS 100
-#define UNUSED_FLAGS 0
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_DEVICE_NODE);
 
-/* Semaphore used to check if the last UART transfer was completed. */
+/* Maximum time to wait for a UART transfer to complete before aborting. */
+#define UART_TX_WAIT_TIME_MS 1000
+/* Maximum UART transfer attempts. */
+#define UART_TX_RETRIES 5
+
+/* Semaphore used to synchronize UART transfers. */
 static K_SEM_DEFINE(tx_sem, 1, 1);
+/* Whether last UART transfer was successful. */
+static bool tx_succeeded;
 
+/* Callback to notify the trace library when trace data is processed. */
 static trace_backend_processed_cb trace_processed_callback;
 
-static void uarte_callback(nrfx_uarte_event_t const *event, void *p_context)
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
 {
-	__ASSERT(k_sem_count_get(&tx_sem) == 0,
-			"UART TX semaphore not in use");
-
-	if (event->type == NRFX_UARTE_EVT_ERROR) {
-		LOG_ERR("uarte error 0x%04x", event->data.error.error_mask);
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
 
+	switch (evt->type) {
+	case UART_TX_DONE:
+		tx_succeeded = true;
 		k_sem_give(&tx_sem);
-	}
-
-	if (event->type == NRFX_UARTE_EVT_TX_DONE) {
+		break;
+	case UART_TX_ABORTED:
+		tx_succeeded = false;
```

Comment:
Looking at the documentation, `UART_TX_ABORTED` doesn't necessarily mean that nothing was sent:

`When flow control is enabled, there is a possibility that TX transfer won’t finish in the allotted time. Some data may have been transferred, information about it can be found in event data.`

In this case, the current implementation would try to send the whole chunk again, so part of the data would be sent twice.

Reply:
Thank you for pointing that out. I'll update this :)